{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import scipy.io\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# Define input directory path (change this to the path where your files are located)\n",
    "input_dir = \"/gpfs/home/asun/jin_lab/get/raw_data/k562\"  # Set the appropriate directory path\n",
    "\n",
    "rna_matrix_file = os.path.join(input_dir, \"GSM8780566_K562_DMSO_RNA_matrix.mtx.gz\")\n",
    "rna_barcodes_file = os.path.join(input_dir, \"GSM8780566_K562_DMSO_RNA_1_barcodes.tsv.gz\")\n",
    "rna_features_file_1 = os.path.join(input_dir, \"GSM8780566_K562_DMSO_RNA_1_features.tsv.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HDF5 file created: K562_DMSO_RNA_combined_replicates.h5ad\n"
     ]
    }
   ],
   "source": [
    "# 1. Load the RNA-seq data matrix (MTX format)\n",
    "rna_matrix_1 = scipy.io.mmread(rna_matrix_file_1).tocsc()  # Sparse matrix format\n",
    "\n",
    "# 2. Load the barcodes (cells)\n",
    "rna_barcodes_1 = pd.read_csv(rna_barcodes_file_1, header=None, sep=\"\\t\").squeeze().tolist()\n",
    "\n",
    "# 3. Load the features (genes)\n",
    "rna_features_1 = pd.read_csv(rna_features_file_1, header=None, sep=\"\\t\")\n",
    "\n",
    "rna_features_1.columns = ['gene_id', 'index', 'feature_type', 'chr', 'start', 'end']\n",
    "\n",
    "rna_features_1['interval'] = rna_features_1.apply(lambda row: f\"{row['chr']}:{row['start']}-{row['end']}\", axis=1)\n",
    "\n",
    "# 4. Create the AnnData object for RNA-seq data (cells x features)\n",
    "adata_rna = sc.AnnData(rna_matrix_1.T)  # Transpose to make (cells x features)\n",
    "\n",
    "adata_rna.obs_names = rna_barcodes_1  # Set the barcodes as obs (cells)\n",
    "adata_rna.var_names = rna_features_1['index'].tolist()\n",
    "\n",
    "rna_features_1.set_index('index', inplace=True)\n",
    "\n",
    "# Store the feature metadata (gene_ids, feature_types, genome, interval) in `adata.var`\n",
    "adata_rna.var['gene_id'] = rna_features_1['gene_id']\n",
    "adata_rna.var['feature_type'] = rna_features_1['feature_type']\n",
    "adata_rna.var['genome'] = \"hg38\"\n",
    "adata_rna.var['interval'] = rna_features_1['interval']\n",
    "\n",
    "# 5. Save the AnnData object to an H5AD file (HDF5 format)\n",
    "#output_file = \"K562_DMSO_RNA_combined_replicates.h5ad\"\n",
    "#adata_rna.write(output_file)\n",
    "\n",
    "print(f\"HDF5 file created: {output_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_file = \"K562_DMSO_RNA_rep1.h5ad\"\n",
    "adata_rna.write(output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_rna = sc.AnnData(rna_matrix_1.T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>genome</th>\n",
       "      <th>interval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1:29553-30267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1:36080-36081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1:65418-69055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1:120931-133723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1:91104-91105</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene_id     feature_type genome            interval\n",
       "MIR1302-2HG  ENSG00000243485  Gene Expression   hg38    chr1:29553-30267\n",
       "FAM138A      ENSG00000237613  Gene Expression   hg38    chr1:36080-36081\n",
       "OR4F5        ENSG00000186092  Gene Expression   hg38    chr1:65418-69055\n",
       "AL627309.1   ENSG00000238009  Gene Expression   hg38  chr1:120931-133723\n",
       "AL627309.3   ENSG00000239945  Gene Expression   hg38    chr1:91104-91105"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_rna.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>index</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>interval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29553</td>\n",
       "      <td>30267</td>\n",
       "      <td>chr1:29553-30267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>36080</td>\n",
       "      <td>36081</td>\n",
       "      <td>chr1:36080-36081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>65418</td>\n",
       "      <td>69055</td>\n",
       "      <td>chr1:65418-69055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>120931</td>\n",
       "      <td>133723</td>\n",
       "      <td>chr1:120931-133723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>91104</td>\n",
       "      <td>91105</td>\n",
       "      <td>chr1:91104-91105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255987</th>\n",
       "      <td>KI270713.1:15882-16673</td>\n",
       "      <td>KI270713.1:15882-16673</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>15882</td>\n",
       "      <td>16673</td>\n",
       "      <td>KI270713.1:15882-16673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255988</th>\n",
       "      <td>KI270713.1:21469-22377</td>\n",
       "      <td>KI270713.1:21469-22377</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>21469</td>\n",
       "      <td>22377</td>\n",
       "      <td>KI270713.1:21469-22377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255989</th>\n",
       "      <td>KI270713.1:26768-27527</td>\n",
       "      <td>KI270713.1:26768-27527</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>26768</td>\n",
       "      <td>27527</td>\n",
       "      <td>KI270713.1:26768-27527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255990</th>\n",
       "      <td>KI270713.1:29577-30473</td>\n",
       "      <td>KI270713.1:29577-30473</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>29577</td>\n",
       "      <td>30473</td>\n",
       "      <td>KI270713.1:29577-30473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255991</th>\n",
       "      <td>KI270713.1:36790-37713</td>\n",
       "      <td>KI270713.1:36790-37713</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>36790</td>\n",
       "      <td>37713</td>\n",
       "      <td>KI270713.1:36790-37713</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>255992 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       gene_id                   index     feature_type  \\\n",
       "0              ENSG00000243485             MIR1302-2HG  Gene Expression   \n",
       "1              ENSG00000237613                 FAM138A  Gene Expression   \n",
       "2              ENSG00000186092                   OR4F5  Gene Expression   \n",
       "3              ENSG00000238009              AL627309.1  Gene Expression   \n",
       "4              ENSG00000239945              AL627309.3  Gene Expression   \n",
       "...                        ...                     ...              ...   \n",
       "255987  KI270713.1:15882-16673  KI270713.1:15882-16673            Peaks   \n",
       "255988  KI270713.1:21469-22377  KI270713.1:21469-22377            Peaks   \n",
       "255989  KI270713.1:26768-27527  KI270713.1:26768-27527            Peaks   \n",
       "255990  KI270713.1:29577-30473  KI270713.1:29577-30473            Peaks   \n",
       "255991  KI270713.1:36790-37713  KI270713.1:36790-37713            Peaks   \n",
       "\n",
       "               chr   start     end                interval  \n",
       "0             chr1   29553   30267        chr1:29553-30267  \n",
       "1             chr1   36080   36081        chr1:36080-36081  \n",
       "2             chr1   65418   69055        chr1:65418-69055  \n",
       "3             chr1  120931  133723      chr1:120931-133723  \n",
       "4             chr1   91104   91105        chr1:91104-91105  \n",
       "...            ...     ...     ...                     ...  \n",
       "255987  KI270713.1   15882   16673  KI270713.1:15882-16673  \n",
       "255988  KI270713.1   21469   22377  KI270713.1:21469-22377  \n",
       "255989  KI270713.1   26768   27527  KI270713.1:26768-27527  \n",
       "255990  KI270713.1   29577   30473  KI270713.1:29577-30473  \n",
       "255991  KI270713.1   36790   37713  KI270713.1:36790-37713  \n",
       "\n",
       "[255992 rows x 7 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_features_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>genome</th>\n",
       "      <th>interval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hg38</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hg38</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hg38</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hg38</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hg38</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id feature_type genome interval\n",
       "MIR1302-2HG     NaN          NaN   hg38      NaN\n",
       "FAM138A         NaN          NaN   hg38      NaN\n",
       "OR4F5           NaN          NaN   hg38      NaN\n",
       "AL627309.1      NaN          NaN   hg38      NaN\n",
       "AL627309.3      NaN          NaN   hg38      NaN"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_rna.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'tolist'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_1587/3063102981.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mrna_barcodes_1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrna_barcodes_file_1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mrna_barcodes_1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniforge3/envs/get/lib/python3.12/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   6314\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mname\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_accessors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6315\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6316\u001b[0m         \u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6317\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6318\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'tolist'"
     ]
    }
   ],
   "source": [
    "rna_barcodes_1 = pd.read_csv(rna_barcodes_file_1, header=None, sep=\"\\t\").tolist()\n",
    "rna_barcodes_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29553</td>\n",
       "      <td>30267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>36080</td>\n",
       "      <td>36081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>65418</td>\n",
       "      <td>69055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>120931</td>\n",
       "      <td>133723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>91104</td>\n",
       "      <td>91105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255987</th>\n",
       "      <td>KI270713.1:15882-16673</td>\n",
       "      <td>KI270713.1:15882-16673</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>15882</td>\n",
       "      <td>16673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255988</th>\n",
       "      <td>KI270713.1:21469-22377</td>\n",
       "      <td>KI270713.1:21469-22377</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>21469</td>\n",
       "      <td>22377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255989</th>\n",
       "      <td>KI270713.1:26768-27527</td>\n",
       "      <td>KI270713.1:26768-27527</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>26768</td>\n",
       "      <td>27527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255990</th>\n",
       "      <td>KI270713.1:29577-30473</td>\n",
       "      <td>KI270713.1:29577-30473</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>29577</td>\n",
       "      <td>30473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255991</th>\n",
       "      <td>KI270713.1:36790-37713</td>\n",
       "      <td>KI270713.1:36790-37713</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>36790</td>\n",
       "      <td>37713</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>255992 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             0                       1                2  \\\n",
       "0              ENSG00000243485             MIR1302-2HG  Gene Expression   \n",
       "1              ENSG00000237613                 FAM138A  Gene Expression   \n",
       "2              ENSG00000186092                   OR4F5  Gene Expression   \n",
       "3              ENSG00000238009              AL627309.1  Gene Expression   \n",
       "4              ENSG00000239945              AL627309.3  Gene Expression   \n",
       "...                        ...                     ...              ...   \n",
       "255987  KI270713.1:15882-16673  KI270713.1:15882-16673            Peaks   \n",
       "255988  KI270713.1:21469-22377  KI270713.1:21469-22377            Peaks   \n",
       "255989  KI270713.1:26768-27527  KI270713.1:26768-27527            Peaks   \n",
       "255990  KI270713.1:29577-30473  KI270713.1:29577-30473            Peaks   \n",
       "255991  KI270713.1:36790-37713  KI270713.1:36790-37713            Peaks   \n",
       "\n",
       "                 3       4       5  \n",
       "0             chr1   29553   30267  \n",
       "1             chr1   36080   36081  \n",
       "2             chr1   65418   69055  \n",
       "3             chr1  120931  133723  \n",
       "4             chr1   91104   91105  \n",
       "...            ...     ...     ...  \n",
       "255987  KI270713.1   15882   16673  \n",
       "255988  KI270713.1   21469   22377  \n",
       "255989  KI270713.1   26768   27527  \n",
       "255990  KI270713.1   29577   30473  \n",
       "255991  KI270713.1   36790   37713  \n",
       "\n",
       "[255992 rows x 6 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_features_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import scipy.io\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# Define input directory path (change this to the path where your files are located)\n",
    "input_dir = \"/gpfs/home/asun/jin_lab/get/raw_data/k562\"  # Set the appropriate directory path\n",
    "\n",
    "rna_matrix_file_2 = os.path.join(input_dir, \"GSM8780567_K562_DMSO_RNA_2_matrix.mtx.gz\")\n",
    "rna_barcodes_file_2 = os.path.join(input_dir, \"GSM8780567_K562_DMSO_RNA_2_barcodes.tsv.gz\")\n",
    "rna_features_file_2 = os.path.join(input_dir, \"GSM8780567_K562_DMSO_RNA_2_features.tsv.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'output_file' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 32\u001b[0m\n\u001b[1;32m     26\u001b[0m adata_rna\u001b[38;5;241m.\u001b[39mvar[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124minterval\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m rna_features_2[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124minterval\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m     28\u001b[0m \u001b[38;5;66;03m# 5. Save the AnnData object to an H5AD file (HDF5 format)\u001b[39;00m\n\u001b[1;32m     29\u001b[0m \u001b[38;5;66;03m#output_file = \"K562_DMSO_RNA_combined_replicates.h5ad\"\u001b[39;00m\n\u001b[1;32m     30\u001b[0m \u001b[38;5;66;03m#adata_rna.write(output_file)\u001b[39;00m\n\u001b[0;32m---> 32\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mHDF5 file created: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[43moutput_file\u001b[49m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'output_file' is not defined"
     ]
    }
   ],
   "source": [
    "# 1. Load the RNA-seq data matrix (MTX format)\n",
    "rna_matrix_2 = scipy.io.mmread(rna_matrix_file_2).tocsc()  # Sparse matrix format\n",
    "\n",
    "# 2. Load the barcodes (cells)\n",
    "rna_barcodes_2 = pd.read_csv(rna_barcodes_file_2, header=None, sep=\"\\t\").squeeze().tolist()\n",
    "\n",
    "# 3. Load the features (genes)\n",
    "rna_features_2 = pd.read_csv(rna_features_file_2, header=None, sep=\"\\t\")\n",
    "\n",
    "rna_features_2.columns = ['gene_id', 'index', 'feature_type', 'chr', 'start', 'end']\n",
    "\n",
    "rna_features_2['interval'] = rna_features_2.apply(lambda row: f\"{row['chr']}:{row['start']}-{row['end']}\", axis=1)\n",
    "\n",
    "# 4. Create the AnnData object for RNA-seq data (cells x features)\n",
    "adata_rna = sc.AnnData(rna_matrix_2.T)  # Transpose to make (cells x features)\n",
    "\n",
    "adata_rna.obs_names = rna_barcodes_2  # Set the barcodes as obs (cells)\n",
    "adata_rna.var_names = rna_features_2['index'].tolist()\n",
    "\n",
    "rna_features_2.set_index('index', inplace=True)\n",
    "\n",
    "# Store the feature metadata (gene_ids, feature_types, genome, interval) in `adata.var`\n",
    "adata_rna.var['gene_id'] = rna_features_2['gene_id']\n",
    "adata_rna.var['feature_type'] = rna_features_2['feature_type']\n",
    "adata_rna.var['genome'] = \"hg38\"\n",
    "adata_rna.var['interval'] = rna_features_2['interval']\n",
    "\n",
    "# 5. Save the AnnData object to an H5AD file (HDF5 format)\n",
    "#output_file = \"K562_DMSO_RNA_combined_replicates.h5ad\"\n",
    "#adata_rna.write(output_file)\n",
    "\n",
    "print(f\"HDF5 file created: {output_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'adata_rna' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43madata_rna\u001b[49m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'adata_rna' is not defined"
     ]
    }
   ],
   "source": [
    "adata_rna"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "get",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
